EEJT-0003/src/main/java/com/epam/esm/giftcertficate/service/GiftCertificateService.java

```diff
@@ -0,0 +1,220 @@
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.domain.*;
+import org.springframework.data.jpa.repository.Modifying;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.EnableTransactionManagement;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.*;
+import java.util.function.Consumer;
+
+@Slf4j
+@Service
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+@EnableTransactionManagement
+public class GiftCertificateService {
+
+    private final GiftCertificateRepository giftCertificateRepository;
+    private final TagService tagService;
+
+    private static final String TAGS = "tags";
+    private static final String NAME = "name";
+    private static final String CREATE_DATE = "createDate";
+    private static final String PERCENT_FOR_SEARCHING_BY_PART = "%";
+
+    @Transactional
+    public GiftCertificate createGiftCertificate(GiftCertificate giftCertificate) {
+        log.info("Service receives gift-certificate for creating");
+        log.info("Transaction has been started");
+
+        log.debug("Checking that gift-certificate has not already exist");
+        if (!giftCertificateRepository.isGiftCertificateExistByName(giftCertificate.getName())) {
+
+            log.debug("Checking tags of gift-certificate");
+            if (giftCertificate.getTags() != null) {
+
+                log.debug("Verify is tag exist");
+                tagService.verifyIsTagsExistWhenCreatingOrUpdatingGiftCertificate(giftCertificate.getTags());
+            }
+            log.info("Transaction has been ended");
+            log.debug("Service returns representation of saved gift-certificate");
+            return giftCertificateRepository.save(giftCertificate);
+        }
+        log.info("Transaction has been ended");
+        log.error("Rollback");
+        log.error("This gift certificate with (name = " + giftCertificate.getName() + ") has already existed");
+        throw new ServerException("this gift certificate with (name = " + giftCertificate.getName() + ") has already existed");
+    }
+
+    public Page<GiftCertificate> getAll(int page, int size) {
+        log.info("Service receives params for getting all");
+
+        PageRequest pageRequest = PageRequest.of(page, size);
+
+        log.debug("Service returns all gift-certificates");
+        return giftCertificateRepository.findAll(pageRequest);
+    }
+
+    public GiftCertificate getOneGiftCertificateById(long id) {
+        log.info("Service receives params for getting gift-certificate by id");
+        log.debug("Service returns one gift-certificate");
+
+        return giftCertificateRepository.findById(id).orElseThrow(
+                () -> {
+                    log.error("there are no gift certificate with (id = " + id + ")");
+                    return new ItemNotFoundException("there are no gift certificate with (id = " + id + ")");
```

Comment:
```suggestion
        return giftCertificateRepository.findById(id).orElseThrow(this::extractedMethod);
```
You can extract it into a private method and remove duplicate code in the following methods

Reply:
i have just one realization with such code , but nice idea